### PR TITLE
Adding fsdp fp16 and bf16 hooks

### PR DESCRIPTION
**Pull request by aovladi**
Recently, `register_comm_hook` was introduced to `FSDP`, which at the moment supports only `NO_SHARD` strategy and has a default `all_reduce` hook implemented. This PR adds two lower precision hooks to an existing default hook.

I've also made slight adjustments to existing implementation of an `all_reduce` hook including:

- `AllReduceState` ->  `DefaultState` , motivation: `AllReduceState` is not specific to `all_reduce`. Gradients' pre- and post-division factors are also useful for other hooks, that require pre- and post-division, e.g. fp16_hook and bf16_hook.
- I've put all 3 hooks into `default_hooks.py`

Additionally, `FSDP` supports `MixedPrecision` and, theoretically, it is possible to specify `MixedPrecision` for gradients and attach a lower precision hook to the model. To avoid double-casting, I've added a couple of checks to `fully_sharded_data_parallel`, i.e. casting to precision and back is performed by a lower precision hook only. I think, as a next step, it would be nice to ensure that user can't have both lower precision hook and `MixedPrecision(reduce_dtype=<precision>)` specified, but I am happy to discuss this and adjust current implementation.

As a test, I create two models: one with a lower precision hook and one with a `MixedPrecision(reduce_dtype=<precision>)` specified, perform one forward/backward and optimizer step and compare gradients.

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_comm_hooks.py [hunk 1]
```diff
@@ -286,6 +291,82 @@ def test_registering_hook_submodules(
...
+        fsdp_with_mp = self._init_model(
+            Net(has_wrapping=has_wrapping, sharding_strategy=sharding_strategy, mixed_precision=mp_only_grad),
+            sharding_strategy=sharding_strategy,
+            mixed_precision=mp_only_grad
+        )
+
+        optim_hook = torch.optim.SGD(fsdp_with_hook.parameters(), lr=0.1)
+        optim_mp = torch.optim.SGD(fsdp_with_mp.parameters(), lr=0.1)
+
+        in_data = torch.rand(16, 8).cuda()
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_comm_hooks.py:315]: nit: torch.rand(16, 8, device='cuda')

File: test/distributed/fsdp/test_fsdp_comm_hooks.py [hunk 2]
```diff
@@ -286,6 +291,82 @@ def test_registering_hook_submodules(
...
+        fsdp_with_hook.train()
+        fsdp_with_mp.train()
+        optim_mp.zero_grad()
+        optim_hook.zero_grad()
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_comm_hooks.py:319]: I don't think zero_grad does anything right, since we haven't accumulated any grads yet?

File: test/distributed/fsdp/test_fsdp_comm_hooks.py [hunk 3]
```diff
@@ -286,6 +291,82 @@ def test_registering_hook_submodules(
...
+        loss_hook.backward()
+        # Make sure grads were cast to the parameter's precision
+        self.assertEqual(fsdp_with_hook.params[0].dtype, state.parameter_type)
+        loss_mp.backward()
+        optim_hook.step()
+        optim_mp.step()
+
+        dist.barrier()
+
+        for hook_params, mp_params in zip(fsdp_with_hook.parameters(), fsdp_with_mp.parameters()):
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_comm_hooks.py:331]: nit: `hook_param, mp_param` since it is a single param

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
+import torch
+import torch.distributed as dist
+from torch.distributed import distributed_c10d
+
+
+class DefaultState(object):
```
Comment (rohan-varma) [inline torch/distributed/algorithms/_comm_hooks/default_hooks.py:6]: Great job to refactor into DefaultState and make LowPrecisionState from DefaultState!

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        state (DefaultState): State information, configures pre- and post-division factors
+        grad (torch.Tensor): A gradient for the local batch that needs to be communicated across ranks.
+    """
+    if state.gradient_predivide_factor > 1:
+        grad.div_(state.gradient_predivide_factor)
+    dist.all_reduce(grad, group=state.process_group)
+    if state.gradient_postdivide_factor > 1:
+        grad.div_(state.gradient_postdivide_factor)
+
+def fp16_compress_hook(state: LowPrecisionState, grad: torch.Tensor):
```
Comment (rohan-varma) [inline torch/distributed/algorithms/_comm_hooks/default_hooks.py:94]: I think we can consolidate fp16 and bf16 hooks, such as the following:


```
def reduced_precision_hook(prec, state, grad):
    grad.data = grad.data.to(prec)
    allreduce_hook(state, grad)
    _decompress(state, grad)
```

```
def fp16_compress_hook(state, grad):
    fp16_hook = functools.partial(reduced_precision_hook, prec=torch.float16)
    return fp16_hook(state, grad)
```

and same for bfloat

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+class DefaultState(object):
+    r"""
+    Stores state needed to perform the default ``all_reduce`` algorithm
+    within a communication hook.
+
+    Args:
+        process_group (ProcessGroup): The process group to be used for all-reduce.
+        world_size (int): The number of workers in a process group.
+            Determined based on a ``process_group``.
+        gradient_predivide_factor (float): A factor for gradients' pre-division.
```
Comment (rohan-varma) [inline torch/distributed/algorithms/_comm_hooks/default_hooks.py:15]: I guess these aren't actually args into DefaultState, so we can remove them from docs?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 7]
```diff
@@ -3000,8 +3015,13 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     # would thus be in this reduced precision, or
                     # 2) parameters did not have precision reduced, but grads
                     # had reduced precision for communication.
+                    # If a lower precision hook is registered, gradients are casted
+                    # back by the hook.
+                    # However, if a low precision hook is attached to the model.
+                    # casting happens inside the hook.
                     if (
-                        self._mixed_precision_enabled_for_params() or self._mixed_precision_enabled_for_reduce()
+                        not self._low_precision_hook_enabled() and
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3023]: We should think about how to refactor this, such as, if reduce_dtype is available, then just register a reduced precision communication hook, but that requires a bit more flushing out. Might be good to file a follow up issue.

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    # setting two factors `self.gradient_predivide_factor`
+    # and `self.gradient_postdivide_factor` to avoid underflow and overflow
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
+            factor *= 2
+        return float(factor)
+
+class LowPrecisionState(DefaultState):
+    r"""
```
Comment (rohan-varma) [inline torch/distributed/algorithms/_comm_hooks/default_hooks.py:46]: Mention that it casts gradients back to the original parameter precision specified by `parameter_type` in the doc?

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    """
+
+    __slots__ = [
+        "parameter_type",
+    ]
+
+    def __init__(
+        self,
+        process_group,
+        parameter_type,
```
Comment (rohan-varma) [inline torch/distributed/algorithms/_comm_hooks/default_hooks.py:62]: let's just default this to torch.float32, which is the usual case so user doesn't have to specify everytime?

**Review by rohan-varma — COMMENTED** [review]
Great work! A couple of suggestions

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py (hunk 5 above)
Comment (aovladi) [reply torch/distributed/algorithms/_comm_hooks/default_hooks.py:94]: Thanks for the example!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 7 above)
Comment (aovladi) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:3077]: I will create a follow-up issue after this PR is landed

File: torch/distributed/algorithms/_comm_hooks/default_hooks.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    approach that casts ``grad`` to half-precision floating-point format (``torch.float16``).
+    It also averages gradients by ``world_size`` in two steps: first it pre-divides gradients by a
+    ``state.predivide_factor``, and after an allreduce step gradients are averaged by a ``state.postdivide_factor``.
+    Onse post-division is done, compressed gradients are casted back to parameters' precision.
+
+    Args:
+        state (DefaultState): State information, configures pre- and post-division factors
+        grad (torch.Tensor): A gradient for the local batch that needs to be communicated across ranks in a lower precision.
+    """
+    bp16_hook = functools.partial(lower_precision_hook, torch.bfloat16)
```
Comment (rohan-varma) [inline torch/distributed/algorithms/_comm_hooks/default_hooks.py:124]: nit: bf16_hook

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 7 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:3086]: We check `not self._low_precision_hook_enabled` in L2980, so this check should always be true within right?

**Review by rohan-varma — APPROVED** [review]
LGTM, one non-minor comment, pls take a look at it

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 7 above)
Comment (aovladi) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:3086]: @rohan-varma right, thanks for catching this

Comment (aovladi) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:3130]: @rohan-varma after looking into this: first check on now L3024 is for pre-casting gradients for all cases. This line is for re-casting gradients back into full-precision for `NO_SHARD`. This does not belong to the body of L3023 if, so this check is still needed. Otherwise, `comm_hook` will cast grads back and will occupy memory for `orig_param_grad_data` and after comm_hook is done, without this check here, it will attempt a second re-casting

Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:3130]: Sounds good
